### PR TITLE
node_modules: switch from deep_equal to dequal

### LINKED ISSUE #20646: Consider switching to dequal
This doesn't help us a lot but dequal is a 14 kB library versus deep-equal which is 90 kB, however has 16 dependencies while dequal has zero. So in the end it might help if other deps drop those dependencies as well.

https://www.npmjs.com/package/dequal

### PR DESCRIPTION
**Pull request by jelly**
Fixes #20646

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "@patternfly/react-styles": "5.3.1",
     "@patternfly/react-table": "5.3.3",
     "@patternfly/react-tokens": "5.3.1",
-    "deep-equal": "2.2.3",
+    "dequal": "2.0.0",
```
Comment (martinpitt) [inline package.json:13]: IMHO dequal/lite is enough. The size difference is marginal, but it also does less checks and thus is faster. We don't need to compare ArrayBuffers and such, this is mostly just about comparing simple dicts.

Not a blocker of course, if this works, let's go!

**Comment by martinpitt** [comment]
Nice, works! Except for this missing types thing:

> pkg/lib/hooks.ts(22,24): error TS7016: Could not find a declaration file for module 'dequal/lite'. '/work/make-checkout-workdir/node_modules/dequal/lite/index.js' implicitly has an 'any' type.

**Comment by jelly** [comment]
> Nice, works! Except for this missing types thing:
> 
> > pkg/lib/hooks.ts(22,24): error TS7016: Could not find a declaration file for module 'dequal/lite'. '/work/make-checkout-workdir/node_modules/dequal/lite/index.js' implicitly has an 'any' type.

Apparently it is typed but in (`dequal` and not `dequal/lite`) , but it is not working obviously. This is not worthwhile to spend time on imo we should drop this user as no one uses useDeepEqualMemo

**Comment by jelly** [comment]
> Nice, works! Except for this missing types thing:
> 
> > pkg/lib/hooks.ts(22,24): error TS7016: Could not find a declaration file for module 'dequal/lite'. '/work/make-checkout-workdir/node_modules/dequal/lite/index.js' implicitly has an 'any' type.

Nothing I can do about it, what works is:
```
import { dequal } from 'dequal';
```

**Comment by martinpitt** [comment]
Dropping `useDeepEqualMemo` seems fine to me, if that helps. Two other places use `deep_equal` directly. Also, this function is so simple, it'd also be fine to just copy it to pkg/lib/utils.tsx.

File: pkg/lib/hooks.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 import cockpit from 'cockpit';
 import { useState, useEffect, useRef, useReducer } from 'react';
-import deep_equal from "deep-equal";
+import { dequal } from 'dequal';
```
Comment (martinpitt) [inline pkg/lib/hooks.ts:22]: Can we use /lite here as well? IMHO it'd be better to just have *one* function in the bundle, not two. We don't need the more "exotic" data types that full dequal supports.

Comment (jelly) [reply pkg/lib/hooks.ts:22]: Well it messed up typescript and I am kinda done with dealing with typescript but yes *ugh* blocker it is then.

File: package.json (hunk 1 above)
Comment (jelly) [reply package.json]: Is that a separate module which I can put in `package.json`? Uh

Comment (jelly) [reply package.json]:
```
[jelle@t14s][~/projects/cockpit/dequal-tryout]%ls node_modules/dequal/
dist/  index.d.ts  license  lite/  package.json  readme.md
```

Is what we have now?

Comment (jelly) [reply package.json]:
```
npm error Invalid package name "dequal/lite" of package "dequal/lite@2.0.3": name can only contain URL-friendly characters.
```
So it seems we already have `lite` :)

File: pkg/lib/hooks.ts (hunk 2 above)
Comment (jelly) [reply pkg/lib/hooks.ts]: By some miracle, typescript is now happy, so I am happy.

File: package.json (hunk 1 above)
Comment (martinpitt) [reply package.json]: Yes, don't confuse the npm package name (dequal) with the `import` module names (dequal or dequal/lite). We "have" it in the sense of "transitive build dependency", but we don't have it yet in the runtime bundles in dist/.

File: pkg/lib/hooks.ts (hunk 2 above)
Comment (martinpitt) [reply pkg/lib/hooks.ts]: Sorry for the confusion -- I didn't mean "typescript" here. This is just a decision whether or not you need the additional features of "dequal" for comparing e.g. `ArrayBuffer` or `DataView` objects (that's what I meant with "exotic data types")

File: package.json (hunk 1 above)
Comment (martinpitt) [reply package.json]: Right, and I did that comment against the wrong line. Anyway, all good now!

**Review by martinpitt — APPROVED** [review]
Happy bots, happy pitti! Thanks!

**Comment by jelly** [comment]
Note, `cockpit-machines` depends on `deep-equal` so needs to be fixed.

**Pull request merged**